src/expr/src/expr/array.rs

```diff
@@ -0,0 +1,214 @@
+        })
+    }
+}
+
+impl Expression for ArrayToStringExpression {
+    fn return_type(&self) -> DataType {
+        DataType::Varchar
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let list_array = self.array.eval_checked(input)?;
+        let list_array = list_array.as_list();
+
+        let delim_array = self.delimiter.eval_checked(input)?;
+        let delim_array = delim_array.as_utf8();
+
+        let null_string_array = if let Some(expr) = &self.null_string {
+            let null_string_array = expr.eval_checked(input)?;
+            Some(null_string_array)
+        } else {
+            None
+        };
+
+        let mut output = Utf8ArrayBuilder::with_meta(input.capacity(), ArrayMeta::Simple);
+
+        for (i, vis) in input.vis().iter().enumerate() {
+            if !vis {
+                output.append_null();
+            } else {
+                let array = list_array.value_at(i);
+                let delim = delim_array.value_at(i);
+                let null_string = if let Some(a) = &null_string_array {
+                    a.as_utf8().value_at(i)
+                } else {
+                    None
+                };
+
+                if let Some(array) = array && let Some(delim) = delim {
+                    let mut writer = output.writer().begin();
+                    if let Some(null_string) = null_string {
+                        evaluate_with_nulls(array, delim, null_string, &mut writer);
+                    } else {
+                        evaluate(array, delim, &mut writer);
+                    }
+                    writer.finish();
+                } else {
+                    output.append_null();
+                }
+            }
+        }
+        Ok(Arc::new(output.finish().into()))
+    }
+
+    fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
+        let array = self.array.eval_row(input)?;
+        let delimiter = self.delimiter.eval_row(input)?;
+
+        let result = if let Some(array) = array && let Some(delimiter) = delimiter && let Some(e) = &self.null_string {
+            let null_string = e.eval_row(input)?;
+            let mut writer = String::new();
+            if let Some(null_string) = null_string {
+                evaluate_with_nulls(array.as_scalar_ref_impl().into_list(), delimiter.as_utf8(), null_string.as_utf8(), &mut writer);
+            } else {
+                evaluate(array.as_scalar_ref_impl().into_list(), delimiter.as_utf8(), &mut writer);
+            }
+            Some(writer)
+        } else {
+            None
+        };
+        Ok(result.map(|r| r.into()))
+    }
+}
+
+fn evaluate(array: ListRef<'_>, delimiter: &str, mut writer: &mut dyn Write) {
+    let mut first = true;
+    for element in array.values_ref().iter().flat_map(|f| f.iter()) {
+        if !first {
+            write!(writer, "{}", delimiter).unwrap();
+        } else {
+            first = false;
+        }
+        element.write(&mut writer).unwrap();
```

Comment:
We should use `write_with_type` rather than `write`, which treats `ScalarImpl::Int64` as `bigint` even when it is actually `timestamptz`.

Reply:
Thanks! What a trap!